File: apps/api/app/infra/observability/sentry.py
"""Sentry scaffold with secret-redacting `before_send`.

Behavior:
- `init_sentry(dsn, environment)` is a no-op when `dsn` is empty. Production
  deployments set `SENTRY_DSN` in `.env`; dev env leaves it blank so no
  outbound traffic is ever emitted by default.
- `before_send(event, hint)` walks the event structure and masks anything
  that smells like an LLM key, an ASR key, or the candidate's free-text
  answer. This is a hard invariant — see phase3-constraints §C and P5.5.

Redacted targets (key match, case-insensitive, keys normalized via `lower()`):
- `x-llm-config` (header carrying base64 LLMConfig on every request)
- any key containing `api_key` / `subscription` / `llm_config` / `azure_speech`
- `authorization`, `bearer`, `token`, `password`           (generic auth)
- `answer`                                                  (InterviewTurn PII)

Value targets:
- `bytes` values are replaced with `"<bytes len=N>"` so raw opus audio
  chunks never reach Sentry intact.
"""

from __future__ import annotations

from typing import Any

import sentry_sdk


_REDACTED = "<redacted>"

_SECRET_EXACT_KEYS: frozenset[str] = frozenset(
    {
        "authorization",
        "bearer",
        "token",
        "password",
        "apikey",
        "answer",
    }
)

# Substring probes kept generic enough not to reference the literal env
# name or the Azure SDK's field name — those live only in
# app/infra/asr/{config,factory,azure_backend}.py, enforced by the
# meta-test at tests/infra/test_asr.py::test_no_secret_token_leakage_in_app_tree.
_SECRET_KEY_SUBSTRINGS: tuple[str, ...] = (
    "llm-config",
    "llm_config",
    "azure_speech",
    "api_key",
    "subscription",
)


def _is_secret_key(key: object) -> bool:
    if not isinstance(key, str):
        return False
    normalized = key.lower()
    if normalized in _SECRET_EXACT_KEYS:
        return True
    for probe in _SECRET_KEY_SUBSTRINGS:
        if probe in normalized:
            return True
    return False


def _redact_node(node: Any) -> Any:
    if isinstance(node, dict):
        return {
            key: (
                _REDACTED
                if _is_secret_key(key)
                else _redact_node(value)
            )
            for key, value in node.items()
        }
    if isinstance(node, list):
        return [_redact_node(item) for item in node]
    if isinstance(node, tuple):
        return tuple(_redact_node(item) for item in node)
    if isinstance(node, (bytes, bytearray, memoryview)):
        length = len(bytes(node))
        return f"<bytes len={length}>"
    return node


def before_send(event: dict[str, Any], hint: dict[str, Any] | None = None) -> dict[str, Any]:
    """Sentry `before_send` hook. Returns the scrubbed event; never drops it."""
    _ = hint  # reserved for future use (e.g. attaching span context safely)
    return _redact_node(event)


def init_sentry(dsn: str, environment: str) -> bool:
    """Initialize Sentry if a non-empty DSN is provided.

    Returns True when the SDK was initialized, False when skipped. Tests
    rely on the False return to assert zero overhead in dev.
    """
    if not dsn:
        return False

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        traces_sample_rate=0.1,
        send_default_pii=False,
        before_send=before_send,
    )
    return True

File: apps/desktop/src/lib/sentry.ts
import * as Sentry from "@sentry/react";

/**
 * Sentry scaffold with secret-redacting `beforeSend`.
 *
 * Mirrors the backend contract at apps/api/app/infra/observability/sentry.py:
 * when the DSN is empty (default), initSentry is a no-op and no network
 * traffic is emitted. Production builds wire VITE_SENTRY_DSN via .env so
 * flipping it activates the scaffold end-to-end.
 *
 * Redaction targets (Phase 5 §P5.5, extends Phase 3 constraint §A1 + §C):
 *   - X-LLM-Config header value (base64 LLMConfig)
 *   - AZURE_SPEECH_KEY / AZURE_SPEECH_REGION env values
 *   - Raw audio Blob / ArrayBuffer bytes
 *   - InterviewTurn.answer free-text content (candidate PII)
 *
 * Matching is case-insensitive on the key; value types are inspected for
 * binary blobs so opus chunks never land in breadcrumbs intact.
 */

const REDACTED = "<redacted>";

// Exact-match keys. Must stay in sync with apps/api/app/infra/observability/sentry.py.
const SECRET_EXACT_KEYS = new Set<string>([
  "authorization",
  "bearer",
  "token",
  "password",
  "apikey",
  "answer",
]);

// Substring probes — kept generic so the literal env name / Azure SDK
// field name only live inside apps/api/app/infra/asr/.
const SECRET_KEY_SUBSTRINGS = [
  "llm-config",
  "llm_config",
  "azure_speech",
  "api_key",
  "subscription",
];

function isSecretKey(key: string): boolean {
  const normalized = key.toLowerCase();
  if (SECRET_EXACT_KEYS.has(normalized)) return true;
  for (const probe of SECRET_KEY_SUBSTRINGS) {
    if (normalized.includes(probe)) return true;
  }
  return false;
}

function redactNode(node: unknown): unknown {
  if (node === null || node === undefined) return node;
  if (typeof node === "string" || typeof node === "number" || typeof node === "boolean") {
    return node;
  }
  if (node instanceof ArrayBuffer) {
    return `<bytes len=${node.byteLength}>`;
  }
  if (typeof Blob !== "undefined" && node instanceof Blob) {
    return `<blob size=${node.size} type=${node.type || "unknown"}>`;
  }
  if (Array.isArray(node)) {
    return node.map((item) => redactNode(item));
  }
  if (typeof node === "object") {
    const out: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(node as Record<string, unknown>)) {
      out[key] = isSecretKey(key) ? REDACTED : redactNode(value);
    }
    return out;
  }
  return node;
}

export function initSentry(dsn: string, environment: string): boolean {
  if (!dsn) return false;

  Sentry.init({
    dsn,
    environment,
    integrations: [Sentry.browserTracingIntegration()],
    tracesSampleRate: 0.1,
    sendDefaultPii: false,
    beforeSend(event) {
      return redactNode(event) as Sentry.ErrorEvent;
    },
    beforeBreadcrumb(breadcrumb) {
      return redactNode(breadcrumb) as Sentry.Breadcrumb;
    },
  });
  return true;
}

// Exported for unit tests and for ad-hoc callers that want to scrub a
// payload before logging it locally.
export const __redactForTests = redactNode;
